pkg/cmd/pulumi/crypto.go

```diff
@@ -23,51 +24,108 @@ import (
-	sm, needsSave, err := getStackSecretsManager(s, ps, nil)
+// A stackSecretsManagerLoader provides methods for loading secrets managers and
+// their encrypters and decrypters for a given stack and project stack. A loader
+// encapsulates the logic for determining which secrets manager to use based on
+// a given configuration, such as whether or not to fallback to the stack state
+// if there is no secrets manager configured in the project stack.
+type stackSecretsManagerLoader struct {
+	// True if the loader should fallback to the stack state if there is no
+	// secrets manager configured in the project stack.
+	FallbackToState bool
+}
+
+// Creates a new stack secrets manager loader from the environment.
+func newStackSecretsManagerLoaderFromEnv() stackSecretsManagerLoader {
+	return stackSecretsManagerLoader{
+		FallbackToState: env.FallbackToStateSecretsManager.Value(),
+	}
+}
+
+// Returns a decrypter for the given stack and project stack.
+func (l *stackSecretsManagerLoader) getDecrypter(
+	ctx context.Context,
+	s backend.Stack,
+	ps *workspace.ProjectStack,
+) (config.Decrypter, bool, error) {
+	sm, needsSave, err := l.getSecretsManager(ctx, s, ps)
 	if err != nil {
 		return nil, false, err
 	}
 
-	enc, err := sm.Encrypter()
+	dec, err := sm.Decrypter()
 	if err != nil {
 		return nil, needsSave, err
 	}
-	return enc, needsSave, nil
+	return dec, needsSave, nil
 }
 
-func getStackDecrypter(s backend.Stack, ps *workspace.ProjectStack) (config.Decrypter, bool, error) {
-	sm, needsSave, err := getStackSecretsManager(s, ps, nil)
+// Returns an encrypter for the given stack and project stack.
+func (l *stackSecretsManagerLoader) getEncrypter(
+	ctx context.Context,
+	s backend.Stack,
+	ps *workspace.ProjectStack,
+) (config.Encrypter, bool, error) {
+	sm, needsSave, err := l.getSecretsManager(ctx, s, ps)
 	if err != nil {
 		return nil, false, err
 	}
 
-	dec, err := sm.Decrypter()
+	enc, err := sm.Encrypter()
 	if err != nil {
 		return nil, needsSave, err
 	}
-	return dec, needsSave, nil
+	return enc, needsSave, nil
 }
 
-func getStackSecretsManager(
-	s backend.Stack, ps *workspace.ProjectStack, fallbackManager secrets.Manager,
+// Returns a secrets manager for the given stack and project stack.
+func (l *stackSecretsManagerLoader) getSecretsManager(
+	ctx context.Context,
+	s backend.Stack,
+	ps *workspace.ProjectStack,
 ) (secrets.Manager, bool, error) {
 	oldConfig := deepcopy.Copy(ps).(*workspace.ProjectStack)
 
 	var sm secrets.Manager
 	var err error
 	if ps.SecretsProvider != passphrase.Type && ps.SecretsProvider != "default" && ps.SecretsProvider != "" {
 		sm, err = cloud.NewCloudSecretsManager(
-			ps, ps.SecretsProvider, false /* rotateSecretsProvider */)
+			ps,
+			ps.SecretsProvider,
+			false, /* rotateSecretsProvider */
+		)
 	} else if ps.EncryptionSalt != "" {
 		sm, err = passphrase.NewPromptingPassphraseSecretsManager(
-			ps, false /* rotateSecretsProvider */)
+			ps,
+			false, /* rotateSecretsProvider */
+		)
```

Comment:
For future reference, these changes make PRs noisier and a lot harder to review. Imo they should be kept either to a separate refactoring PR, or just avoided.  There seem to be some similar instances above (though they may have been necessary to get around the `lll` linter)